bitpacker/src/bitpacker.rs

```diff
@@ -87,28 +91,40 @@ impl BitUnpacker {
 
     #[inline]
     pub fn get(&self, idx: u32, data: &[u8]) -> u64 {
-        if self.num_bits == 0 {
-            return 0u64;
-        }
-        let addr_in_bits = idx * self.num_bits as u32;
+        let addr_in_bits = idx * self.num_bits;
```

Comment:
I was wondering if dynamic dispatch on a compile time num_bits for all 64 variants would be a benefit performance

Reply:
From the usage sites, it looks like the compiler will have a hard time inferring a constant, hence I would suspect that it could help. But I suspect it would also need some kind of type registry to provide access to `&'static dyn` without boxing?

Finally, if the number of actually used instance is really small, it might be better to use "dictionary dispatch", e.g. have a `match` on the bit width and branch into compile-time specializations with known bit widths, c.f. https://lwn.net/Articles/815908/

Reply:
Not sure if what I meant was clear, when opening you'd get a typed version 

```rust
pub fn new(num_bits: u8) -> Arc<dyn BitUnpackerTrait>{
    match num_bits {
        1 => Arc::new(BitUnpacker::<1>),
        2 => Arc::new(BitUnpacker::<2>),
        ...
    }
}
```

That would limit inlining and we have a dynamic dispatch, but it would allow optimization in the `get` method and reduce the number of cycles there.

https://godbolt.org/z/ch4Wsjaf6

A `match` would also be an alternative worth looking into.

With specialized versions, we don't need the padding anymore, since we can read a variable number of bytes, depending on `num_bits`.

Reply:
I suspect a regression but who knows.
That's not a bad idea., at least it could help removing the 0 special case if there are no regressions...
and with the bit=1 case we could also use it to build an adhoc column value for booleans.